src/Storage/FileStorage.php

```diff
@@ -1,476 +1,365 @@
+
+    /** @var bool */
+    protected $readOnly;
+
+    /** @var FileHandlerPhp */
+    protected $fileHandler;
+
+    /**
+     * @throws ParameterException
+     * @throws StorageException
+     */
+    public function __construct(IFileHandler $fileHandler, string $path, bool $readOnly = false)
+    {
+        $this->fileHandler = $fileHandler;
+        $this->readOnly    = $readOnly;
+
+        $this->setPath($path);
+    }
+
+    /**
+     * @throws ParameterException
+     * @throws StorageException
+     */
+    public function setPath(string $path)
+    {
+        if (empty($path)) {
+            throw new ParameterException('Path has to be set');
+        }
+
+        if (substr($path, -1, 1) !== DIRECTORY_SEPARATOR) {
+            $path .= DIRECTORY_SEPARATOR;
+        }
+
+        $this->path = $path;
+
+        $this->initAndValidatePath();
+    }
+
+    /**
+     * @throws StorageException
+     */
+    protected function initAndValidatePath()
+    {
+        if (!$this->fileHandler->checkIsPathExists($this->path)) {
+            try {
+                $this->fileHandler->makeDirectory($this->path, ($this->fileModeOctal | 0111), true);
+            } catch (FileException $e) {
+                throw new StorageException('Can not create directory for FileStorage: ' . $this->path, 0, $e);
+            }
+        } elseif (!$this->fileHandler->checkIsDirectory(rtrim($this->path, '/'))) {
+            throw new StorageException('Path is not a directory for FileStorage: ' . $this->path);
+        }
+
+        if (!$this->readOnly && !$this->fileHandler->checkIsWritable($this->path)) {
+            throw new StorageException('Path is not writable for FileStorage: ' . $this->path);
+        }
+    }
+
+    /**
+     * Returns the full path for the specified filename
+     *
+     * @throws StorageException
+     */
+    protected function getFullPath(string $fileName): string
+    {
+        $fileName = $this->filenamePrefix . $fileName . $this->filenameSuffix;
+        if ($this->hashKey) {
+            $fileName = md5($fileName);
+        }
+
+        if (!preg_match('/^[-_.a-zA-Z0-9]+$/', $fileName)) {
+            throw new StorageException('Invalid filename: ' . $fileName);
+        }
+        return $this->path . $fileName;
+    }
+
+    public function set(string $key, $data, int $ttlInSeconds = 0): void
+    {
+        $this->protectWhenReadOnly();
+
+        $fullPath = $this->getFullPath($key);
+
+        try {
+            $item = (new Storage(Storage::METHOD_SET, $key, $data));
+
+            $this->fileHandler->write($fullPath, $this->prepareDataToStore($data, $ttlInSeconds));
+
+            $item->setFinished();
+            $this->getDebugDataHandlerRegistry()->addStorage($item);
+        } catch (FileException $e) {
+            throw new StorageException('Unable to write data to FileStorage (file: ' . $fullPath . ' )', 0, $e);
+        }
+
+        // Disable potential warnings if unit testing with vfsStream
```

Comment:
BTW we should probably replace vfsStream too... Flysystem may be a good option for file handling, not sure if it covers everything we need it for though...

Reply:
To be honest I left this here because I don't know yet why would vfs trigger warnings.
Let's keep this flysystem thing to the end, when we have a workinf FW we can start experiment with these.
Though I'm almost sure it provides everything we need